### ISSUE TITLE
Questions about the memory usage

### ISSUE THREAD

**Issue by RanTaimu** [facebookresearch/maskrcnn-benchmark#329]
## ❓ Questions and Help
Hi, thanks for your wonderful work! I have some questions and I would appreciate it if you could answer me:
1. **The memory usage shown in nvidia-smi is greater than "max mem" in log by a large margin, even if I put iou computation into CPU.**
    I changed the code in "boxlist_ops.py":
    ```python
    # Converting to cpu to save GPU memory
    device = box1.device
    box1 = box1.cpu()
    box2 = box2.cpu()

    lt = torch.max(box1[:, None, :2], box2[:, :2])  # [N,M,2]
    rb = torch.min(box1[:, None, 2:], box2[:, 2:])  # [N,M,2]

    TO_REMOVE = 1

    wh = (rb - lt + TO_REMOVE).clamp(min=0)  # [N,M,2]
    inter = wh[:, :, 0] * wh[:, :, 1]  # [N,M]

    # iou = inter / (area1[:, None] + area2 - inter)
    iou = inter / (area1[:, None].cpu() + area2.cpu() - inter)
    iou = iou.to(device)
    return iou
    ```
     But the memory shown in nvidia-smi is 31G while logged memory is 25G(I just add an extra branch in mask predictor, which box operation is not involved). Is there something I've missed?

2. **Free memory by torch.cuda.empty_cache()**
     Due to the phenomenon in Question 1, I decided to free the cached memory manually. And I invoked "torch.cuda.empty_cache()" after every iteration. And I found that the memory reported by nvidia-smi showing a rather intense fluctuation(between 10+G and 20+G), even if I computed the IoU on CPU. Do you have any idea on it?

3. Is there any advice to save the GPU memory? I need to leave some memory for other users.

**Comment by fmassa** [comment]
Hi,

The discussion in https://github.com/facebookresearch/maskrcnn-benchmark/issues/182#issuecomment-440583483 and https://github.com/facebookresearch/maskrcnn-benchmark/issues/182#issuecomment-442112743 will answer most of your questions I believe.

I don't think there is a way of restricting the max amount of memory that the caching allocator from PyTorch can use, so unfortunately I don't think there is much we can do for now.

Let me know if you have further questions.

**Comment by RanTaimu** [comment]
@fmassa Thanks for your reply. I read that issue carefully before. The only thing I want to know is the second question I asked. Specifically, after I move box operation to CPU using the code above, why the GPU memory usage was still not stable during training? The max mem is 12557 while 30687 shown in nvidia-smi. It's too strange.

**Comment by RanTaimu** [comment]
@fmassa OK, thanks for your help. I'll close this issue.
